File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {


        ListNode dummy = new ListNode(-1);
        dummy.next = head;

        ListNode pointer = head;
        int counter = 0;
        while(pointer != null) {
            counter++;
            pointer = pointer.next;
        }

        // get the difference
        int elementCounter = counter - n;
        pointer = dummy;


        for(int i=0;i<elementCounter;i++) {
            pointer = pointer.next;
        }

        pointer.next = pointer.next.next;
        // loop through the list again 


        return dummy.next;
    }
}

File: Data Structures & Algorithms/middle-of-the-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:

        if not head:
            return None

        # loop through the array and get the length 
        counter = 0 
        curr = head
        while curr:
            counter = counter + 1
            curr = curr.next

        # get the middle element
        middle = counter//2
        new_counter = 0
            
        curr2 = head

        while curr2:
            if new_counter == middle:
                return curr2
            curr2 = curr2.next
            new_counter = new_counter + 1
            
        
        